service/controller/azuremachinepool/handler/migration/resource.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/v7/pkg/controller/context/reconciliationcanceledcontext"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	capzexp "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v5/service/controller/key"
+)
+
+const (
+	// Name is the identifier of the resource.
+	Name = "migration"
+)
+
+type Config struct {
+	CtrlClient client.Client
+	Logger     micrologger.Logger
+}
+
+type Resource struct {
+	ctrlClient client.Client
+	logger     micrologger.Logger
+}
+
+func New(c Config) (*Resource, error) {
+	if c.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", c)
+	}
+	if c.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", c)
+	}
+
+	r := &Resource{
+		ctrlClient: c.CtrlClient,
+		logger:     c.Logger,
+	}
+
+	return r, nil
+}
+
+func (r Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	azureMachinePool, err := key.ToAzureMachinePool(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if !areReferencesUpdated(azureMachinePool) {
+		// Migration from old to new MachinePool is not completed. Cancel
+		// remaining reconciliation.
+		reconciliationcanceledcontext.SetCanceled(ctx)
+		return nil
+	}
+
+	// Now the new AzureMachinePool has been created with Spec copied from old
+	// AzureMachinePool, so here we want to:
+	// 1. Clone Status from old AzureMachinePool
+	// 2. Delete old AzureMachinePool
+
+	// Get old AzureMachinePool
+	namespacedName := types.NamespacedName{
+		Namespace: azureMachinePool.Namespace,
+		Name:      azureMachinePool.Name,
+	}
+	oldAzureMachinePool := &oldcapzexpv1alpha3.AzureMachinePool{}
+	err = r.ctrlClient.Get(ctx, namespacedName, oldAzureMachinePool)
+	if apierrors.IsNotFound(err) {
+		// Old AzureMachinePool not found, so nothing to do here.
+		return nil
+	} else if err != nil {
+		// Migration from old to new AzureMachinePool is not completed
+		// because we still didn't update the status in the new AzureMachinePool,
+		// so we cancel remaining reconciliation.
+		reconciliationcanceledcontext.SetCanceled(ctx)
+		return microerror.Mask(err)
+	}
+
+	// 1. Update status in new AzureMachinePool
+	if isStatusEmpty(azureMachinePool) {
+		err = cloneObject(&oldAzureMachinePool.Status, &azureMachinePool.Status)
+		if err != nil {
+			// Migration from old to new AzureMachinePool is not completed
+			// because we still didn't update the status in the new AzureMachinePool,
+			// so we cancel remaining reconciliation.
+			reconciliationcanceledcontext.SetCanceled(ctx)
+			return microerror.Mask(err)
+		}
+
+		// Update new AzureMachinePool
+		err = r.ctrlClient.Status().Update(ctx, &azureMachinePool)
```

Comment:
How about labels and annotations? where they copied already?

Reply:
If I understand correctly, the handler reconciling the old experimental CRs creates the new CR with the copied `metadata` and `spec` fields. Then this controller/handler reconciling the new CR copies the `status` and updates the new `CR`, then it deletes the old CR. So yeah, I'd say labels and annotations are already present here.

Reply:
> where they copied already?

Yes, they are copied already, Jose described it pretty accurately.